File: LeetCode/merge_two_sorted_lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        if(!l1)
            return l2;
        if(!l2)
            return l1;
        
        ListNode *aux;
        
        if(l1->val < l2->val) {
            aux = new ListNode(l1->val);
            l1 = l1->next;
        }
        else {
            aux = new ListNode(l2->val);
            l2 = l2->next;
        }
        
        ListNode *head = aux;
        
        while(l1 && l2) {
            if(l1->val < l2->val) {
                aux->next = l1;
                aux = aux->next;
                l1 = l1->next;
            }
            else {
                aux->next = l2;
                aux = aux->next;
                l2 = l2->next;
            }
        }
    
        if(l1)
            aux->next = l1;
        else if(l2)
            aux->next = l2;
        
        return head;
    }
};

File: LeetCode/merge_two_sorted_lists.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
    ans := &ListNode{}
    curr1, curr2, currAns := list1, list2, ans

    for curr1 != nil && curr2 != nil {
        if curr1.Val < curr2.Val {
            currAns.Next = &ListNode{
                Val: curr1.Val,
                Next: nil,
            }
            curr1 = curr1.Next
        } else {
            currAns.Next = &ListNode{
                Val: curr2.Val,
                Next: nil,
            }
            curr2 = curr2.Next
        }
        currAns = currAns.Next
    }

    if curr1 != nil {
        currAns.Next = curr1
    }

    if curr2 != nil {
        currAns.Next = curr2
    }

    return ans.Next
}
